### PR TITLE
Nette 3.0

### PR DESCRIPTION
**Pull request by haltuf**
Prepare Kdyby\Doctrine to support Nette 3.0
Necessary prerequisities are merging of my PR's of Kdyby/Annotations and Kdyby/DoctrineCache

### CODE REVIEW

**Comment by haltuf** [comment]
Please, if you find any error in real applications, report me here.
I will thoroughly test this PR against Nette 3.0 real application during next week.

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (enumag) [inline src/Kdyby/Doctrine/DI/OrmExtension.php]: Is `$this->compiler->getContainerBuilder()` different from `$builder`?

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (enumag) [inline src/Kdyby/Doctrine/DI/OrmExtension.php]: wrong indentation

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (enumag) [inline src/Kdyby/Doctrine/DI/OrmExtension.php]: why not chain?

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (enumag) [inline src/Kdyby/Doctrine/DI/OrmExtension.php]:
```suggestion
				->setFactory($originalDef->getFactory());
```

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (enumag) [inline src/Kdyby/Doctrine/DI/OrmExtension.php]: why?

File: src/Kdyby/Doctrine/EntityRepository.php
Comment (enumag) [inline src/Kdyby/Doctrine/EntityRepository.php]: remove completely if impossible to use... also why?

**Review by enumag — CHANGES_REQUESTED** [review]

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (haltuf) [reply src/Kdyby/Doctrine/DI/OrmExtension.php]: I was following the example of Nette itself, as the last three lines are all the methods of object returned by getResultDefinition().
See: https://github.com/nette/application/blob/master/src/Bridges/ApplicationDI/LatteExtension.php#L57:L62

If you find it still unacceptable, I will change all lines to the same indentation.

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (haltuf) [reply src/Kdyby/Doctrine/DI/OrmExtension.php]: Because then it fails on phpstan.
getDefinition() returns ANY Definition. But the method addSetup() is defined only for ServiceDefinition. I have to tell phpstan, that $def is of type ServiceDefinition first and then call ->addSetup.
Otherwise, phpstan returns error.

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (haltuf) [reply src/Kdyby/Doctrine/DI/OrmExtension.php]: phpstan, same problem as above.
If you know a better fix, please suggest.
We can always whitelist this error message in phpstan, of course.

File: src/Kdyby/Doctrine/EntityRepository.php
Comment (haltuf) [reply src/Kdyby/Doctrine/EntityRepository.php]: Kdyby\StrictObjects\Scream is not compatible with Doctrine\ORM\EntityRepository so the tests fail.
It's possible, that it will be in next versions.
See: https://github.com/Kdyby/StrictObjects/blob/master/src/InstanceMethodCallScreamer.php#L18
vs. https://github.com/doctrine/orm/blob/master/lib/Doctrine/ORM/EntityRepository.php#L175

I could also define __call(string $name, array $args) in Kdyby\Doctrine\EntityRepository and then just call parent::__call($name, $args), if you like such a solution more.

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (enumag) [reply src/Kdyby/Doctrine/DI/OrmExtension.php]: Then how about adding this?

```
private function getServiceDefinition(ContainerBuilder $builder, string $name): ServiceDefinition
{
    $definition = $builder->getDefinition($name);
    assert($definition instanceof ServiceDefinition);

	return $definition;
}
```

Comment (haltuf) [reply src/Kdyby/Doctrine/DI/OrmExtension.php]: good idea, fixed

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (haltuf) [reply src/Kdyby/Doctrine/DI/OrmExtension.php]: no, it's same... fixed.

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (enumag) [reply src/Kdyby/Doctrine/DI/OrmExtension.php]: hmm... yeah I'd prefer same indentation, thanks

File: src/Kdyby/Doctrine/EntityRepository.php
Comment (enumag) [reply src/Kdyby/Doctrine/EntityRepository.php]: hmm... just drop Scream I think...

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (enumag) [inline src/Kdyby/Doctrine/DI/OrmExtension.php]: spaces instead of tabs here

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (haltuf) [reply src/Kdyby/Doctrine/DI/OrmExtension.php]: fixed.

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (haltuf) [reply src/Kdyby/Doctrine/DI/OrmExtension.php]: fixed

File: src/Kdyby/Doctrine/EntityRepository.php
Comment (haltuf) [reply src/Kdyby/Doctrine/EntityRepository.php]: dropped

**Comment by haltuf** [comment]
TODO, before merging:
 - removing dependencies on my temporary branches
 - removing direct dependency on Kdyby/Console

File: src/Kdyby/Doctrine/DI/OrmExtension.php
Comment (haltuf) [reply src/Kdyby/Doctrine/DI/OrmExtension.php]: fixed

**Comment by enumag** [comment]
> removing direct dependency on Kdyby/Console

leave that for another PR

**Comment by vojta001** [comment]
In what time frame can we expect this to be merged? Since Nette 2.4 is EOL, this is a real dealbreaker.

**Comment by enumag** [comment]
Currently blocked by https://github.com/Kdyby/Annotations/pull/23.

**Comment by haltuf** [comment]
Sorry for the delay, I'll get to it next week (Mon or Tue).

File: composer.json
Comment (haltuf) [inline composer.json]: Kdyby/Annotations PR has been merged. After new version is released, we still have to get rid of these temporary dependencies (replace haltuf/DoctrineCache and haltuf/Annotations with kdyby/*)

**Comment by enumag** [comment]
New annotations are released. Get rid of the temporary branches here. Let me know if there is something else blocking this PR.

**Comment by haltuf** [comment]
@enumag Did all the tests with fresh versions and found only one problem solved in [Kdyby/Annotations#24](https://github.com/Kdyby/Annotations/pull/24).

After that, I will push the last commit to Kdyby/Doctrine (replacing the temp reference for kdyby/* ^3.0) and I think we have RC ready.

**Comment by enumag** [comment]
@haltuf annotations fixed and released.

**Comment by haltuf** [comment]
@enumag Ok, I think we have RC here. If there are some errors, I can fix them before Sunday. Then, I wont be available in the whole upcoming week.

**Review by enumag — APPROVED** [review]

**Comment by enumag** [comment]
Nice. I'd still like to see if it is possible to drop the kdyby/console dependency. An ability to use contributte/console instead seems important these days.

**Comment by haltuf** [comment]
@enumag I will prepare a new PR for dropping this dependency in September. It will be challenge for me, but at least I'll learn something new :-)

**Comment by simPod** [comment]
nette/finder 3 is not released yet 🤔

**Comment by spernica** [comment]
@simPod Is version 3 really needed? Why not use latest version (2.5.1) - it's Nette 3 compatible.

**Comment by simPod** [comment]
@spernica no idea, I'm not an author of this PR nor I don't know Nette packages well. 

But does not make much sense to me to require nonexistent version 🤔

**Comment by haltuf** [comment]
@simPod @spernica I believe it's my mistake, am little bit behind everything, but will try to make it right. Does it currently prevent you in using/installing this package?

**Comment by simPod** [comment]
It did but I decided not to upgrade anyway, had no time for that. So no worries.

**Comment by spernica** [comment]
@haltuf Thanks for the info guys, I was just confused, because it doesn't make sense to me. 
We are still on Nette 2.4 using older version of Kdyby/Doctrine.

**Comment by Spamercz** [comment]
Regarding console integration, I would suggest check out how it is hanled in Nextras. Something like here
https://github.com/nextras/migrations/blob/master/src/Bridges/NetteDI/MigrationsExtension.php#L91
And to use both implementations (kdyby/contribute) only thing you need i think is add right tag. As defined here https://github.com/Spameri/Elastic/blob/master/src/Config/Elastic.neon#L163 (but not tested, I am using there only kdyby/console) 

Hope this helps at least a bit :)

**Comment by haltuf** [comment]
Thx, I've came to similar conclusions. :-) The process is not so easy, though.
First, kdyby/doctrine uses something called ContainerHelper from kdyby/console, so I have to unwind this dependency first. My final goal is to be able to:
a/ install kdyby doctrine WITHOUT any console
b/ install kdyby doctrine WITH kdyby console
c/ install kdyby doctrine WITH contributte console

Now when I dived into the problem little bit, I believe it could be doable.

**Pull request merged**
